src/modules/account/RadixDataAccountSystem.ts

```diff
@@ -85,38 +91,49 @@ export default class RadixDataAccountSystem implements RadixAccountSystem {
             hid,
             applicationId,
             data: applicationData,
+            signatures,
         }
 
         this.applicationData.get(applicationId).delete(hid)
         this.applicationDataSubject.next(applicationDataUpdate)
     }
 
-    public getApplicationData(
-        applicationId: string
-    ): Observable<RadixApplicationDataUpdate> {
+    public getApplicationData(applicationId: string, addresses: string[]): Observable<RadixApplicationDataUpdate> {
         return Observable.create(
             (observer: Observer<RadixApplicationDataUpdate>) => {
+
                 // Send all old data
                 if (this.applicationData.has(applicationId)) {
                     for (const applicationData of this.applicationData
                         .get(applicationId)
                         .values()) {
-                        const applicationDataUpdate = {
-                            action: 'STORE',
-                            hid: applicationData.hid,
-                            applicationId: applicationId,
-                            data: applicationData
-                        }
 
-                        observer.next(applicationDataUpdate)
+                        if (!addresses 
+                            || addresses.length === 0
+                            || addresses.some(a => 
+                                Object.keys(applicationData.signatures).includes(RadixKeyPair.fromAddress(a).getUID().toString()))) {
```

Comment:
Is this expensive: `RadixKeyPair.fromAddress(a).getUID().toString()`? Is there a point in caching the output?

Reply:
Not cheap for sure, parsing a string address is quite intensive, another good catch

Reply:
Good point, though the cost of `RadixKeyPair.fromAddress(a).getUID().toString()` is almost depreciable as you can see from running the benchmarks in https://github.com/indutny/bn.js/blob/master/benchmarks/index.js.

I think the most expensive methods here are `some` & `includes`, but they basically iterate over the arrays, so the complexity of this line is in the order of `O(n*m)` where `n = length of addresses` and `m = length of applicationData.signatures`.

Reply:
I think the fromAddress bit is the expensive one, you have to hash and base58 decode

Reply:
That link you shared, @murraco, seems very interesting. But how do I use it?

Reply:
@MuncleUscles I didn't think about the `fromAddress` method, I'll calculate that before making the comparison then

@laszlourszuly you can run the benchmarks with the following commands:

```bash
$ git clone https://github.com/indutny/bn.js
$ cd bn.js/benchmarks/
$ npm I && node index.js
```

I'm used node v8 and I recommend you to look into nvm (https://github.com/creationix/nvm) to easily manage/switch different version.